lib/internal/modules/esm/module_map.js

```diff
@@ -1,17 +1,119 @@
 'use strict';
 
-const { kImplicitAssertType } = require('internal/modules/esm/assert');
 const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  ArrayPrototypeSort,
+  JSONStringify,
+  ObjectDefineProperty,
+  ObjectKeys,
   SafeMap,
 } = primordials;
+const { kImplicitAssertType } = require('internal/modules/esm/assert');
+const { setOwnProperty } = require('internal/util');
 let debug = require('internal/util/debuglog').debuglog('esm', (fn) => {
   debug = fn;
 });
 const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
 const { validateString } = require('internal/validators');
 
-// Tracks the state of the loader-level module cache
-class ModuleMap extends SafeMap {
+/**
+ * Cache the results of the `resolve` step of the module resolution and loading process.
+ * Future resolutions of the same input (specifier, parent URL and import assertions)
+ * must return the same result if the first attempt was successful, per
+ * https://tc39.es/ecma262/#sec-HostLoadImportedModule.
+ */
+class ResolveCache extends SafeMap {
+  constructor(i) { super(i); } // eslint-disable-line no-useless-constructor
+
+  /**
+   * Generates the internal serialized cache key and returns it along the actual cache object.
+   *
+   * It is exposed to allow more efficient read and overwrite a cache entry.
+   * @param {string} specifier
+   * @param {string} parentURL
+   * @param {Record<string,string>} importAssertions
+   * @returns {{
+   *   serializedModuleRequest: string,
+   *   internalCache: Record<string, Promise<import('./loader').ModuleExports> | import('./loader').ModuleExports>,
+   * }}
+   */
+  getSerialized(specifier, parentURL, importAssertions) {
+    let internalCache = super.get(parentURL);
+    if (internalCache == null) {
+      super.set(parentURL, internalCache = { __proto__: null });
+    }
```

Comment:
What is `internalCache`? It seems like _another_ key-value map, but one for each `parentURL`? As in, these are all the imports from each module? Why is this a plain object instead of another `SafeMap`?

Reply:
I copied the implementation from `LoadModuleCache`. There might be performance considerations, or the original author found it more readable this way, not sure.

Reply:
We seem to use `SafeMap` rather than plain objects for dictionaries throughout the codebase, so I'd guess that `SafeMap` is more performant?

Reply:
`Map` is HUGELY more performant over a plain object for a large number of rapid mutations